README.md
**yaoui** is a UI kit for LÖVE. If you need a fast way to build half-decent looking UI without having to worry about much then this module is for you. There are options for customization like changing theme colors or even how each UI element looks, but those are not the main problems this kit tries to solve.

<p align="center">
 <a href="https://imgrush.com/cogghKLdAEuv">
  <img border="0" src="http://i.imgur.com/z1h7AUH.jpg">
 </a>
</p>

## Table of Contents

* [Introduction](#introduction)
  * [View](#view)
  * [Stack](#stack)
  * [Flow](#flow)
  * [Accessing elements](#accessing-elements)
* [Elements](#elements)
  * [Button](#button)
  * [Checkbox](#checkbox)
  * [Dropdown](#dropdown)
  * [FlatDropdown](#flatdropdown)
  * [FlatTextinput](#flattextinput)
  * [HorizontalSeparator](#horizontalseparator)
  * [HorizontalSpacing](#horizontalspacing)
  * [IconButton](#iconbutton)
  * [ImageButton](#imagebutton)
  * [Tabs](#tabs)
  * [Textinput](#textinput)
  * [VerticalSeparator](#verticalseparator)
  * [VerticalSpacing](#verticalspacing)
* [Theme](#theme)
  * [Changing the theme](#changing-the-theme)

## Usage

Require the [module](https://github.com/adonaac/yaoui/tree/master/yaoui):

```lua
yui = require 'yaoui'
```

Register it to most of LÖVE's callbacks:

```lua
function love.load()
  yui.UI.registerEvents()
end
```

And update:

```lua
function love.update(dt)
  yui.update({})
end
```

## Introduction

The main idea behind yaoui is the usage of 3 main elements for directing UI layout: **views**, **stacks** and **flows**. The idea is taken from the Ruby [shoes](http://shoesrb.com/) library. Check that link out since it's immediately obvious what stacks and flows do.

### View

Views are effectively windows that can contain a Stack or a Flow. You can specify the position of a view, as well as its margins:

```lua
function love.load()
  yui.UI.registerEvents()
  yui.debug_draw = true

  view = yui.View(50, 50, 400, 300, {
    margin_top = 10,
    margin_left = 10,
    yui.Stack({
      yui.Button({text = 'Some button', hover = 'Button hover'}),
      yui.Button({text = 'Some button', hover = 'Button hover'}),
      yui.Button({text = 'Some button', hover = 'Button hover'}),
    })
  })
end

function love.update(dt)
  yui.update({view})
  view:update(dt)
end

function love.draw()
  view:draw()
end
```

And that will get you something like this (yui.debug_draw is set to true so you can see View and Stack area rectangles):

![ya1](http://i.imgur.com/VBBfpLi.gif)

---

### Stack

Stacks let you stack UI elements together on top of each other. You can set its margins as well as the spacing between each element. You can also make it so that elements are added from the bottom of the Stack instead of from the top.

Continuing from the previous example, this:

```lua
view = yui.View(50, 50, 400, 300, {
  margin_top = 10,
  margin_left = 10,
  yui.Stack({margin_left = 10, margin_top = 10, margin_bottom = 10, margin_right = 10, spacing = 5,
    yui.Button({text = 'Some button', hover = 'Button hover'}),
    yui.Button({text = 'Some button', hover = 'Button hover'}),
    yui.Button({text = 'Some button', hover = 'Button hover'}),

    bottom = {
      yui.Button({text = 'Other button', hover = 'Button hover'}),
      yui.Button({text = 'Other button', hover = 'Button hover'}),
      yui.Button({text = 'Other button', hover = 'Button hover'}),
    }
  })
})
```

Should get you this:

![ya2](http://i.imgur.com/mmQOZVq.gif)

---

### Flow

Flows let you stack UI elements together from left to right. Like with a Stack, you can set a flows margin as well as spacing between each element. You can also start stacking from the right if you want.

```lua
view = yui.View(50, 50, 600, 300, {
  margin_top = 10,
  margin_left = 10,
  yui.Flow({margin_left = 10, margin_top = 10, margin_bottom = 10, margin_right = 10, spacing = 5,
    yui.Button({text = 'Some button', hover = 'Button hover'}),
    yui.Button({text = 'Some button', hover = 'Button hover'}),

    right = {
      yui.Button({text = 'Other button', hover = 'Button hover'}),
      yui.Button({text = 'Other button', hover = 'Button hover'}),
    }
  })
})
```

![ya3](http://i.imgur.com/I4rMyyo.gif)

---

With these 3 main elements you get a good compromise between not having to do any manual placement but also being able to lay your elements out in a way that looks appealing. You can put Flows inside Stacks, Stacks inside Flows, Stacks inside Stacks, etc and have them be nested in whatever way you want. The only limitation is that a View should only have one element and that element should be either a Stack or a Flow (and then inside this Stack/Flow you can do whatever).

### Accessing elements

All elements in the View tree can be accessed from the view. In the examples above, for instance, if we wanted to access the right-most button on the Flow, we'd do:

```lua
view[1].right[2]
```

Here `[1]` refers to the Flow, `right` refers to elements that stack from the right, and `[2]` refers to the second button (right-most). To make thinks easier we can also define `name` attributes for each element, for instance:

```lua
view = yui.View(50, 50, 600, 300, {
  margin_top = 10,
  margin_left = 10,
  yui.Flow({
    name = 'MainFlow',
    margin_left = 10, margin_top = 10, margin_bottom = 10, margin_right = 10, spacing = 5,
    yui.Button({text = 'Some button', hover = 'Button hover'}),
    yui.Button({text = 'Some button', hover = 'Button hover'}),

    right = {
      yui.Button({text = 'Other button', hover = 'Button hover'}),
      yui.Button({name = 'RightMostButton', text = 'Other button', hover = 'Button hover'}),
    }
  })
})
```

And now to get the right-most button we can do:

```lua
view.MainFlow.right.RightMostButton
```

All UI elements created with yaoui can have a `name` attribute attached to them so that they can be accessed in this manner. If they don't have this `name` attribute they can still be accessed via their index position.

## Elements

Here follows a description of all UI elements. For elements where *icons* are available they refer to [Font Awesome Icons](http://fontawesome.io/icons/). All icons can be specified with `icon = fa-icon_name` where `icon_name` is the Font Awesome icon name.

### Button

```lua
yui.Button({text = 'Button text', onClick = function(self) print(1) end})
```

#### Construction options

| Option | Description | Mandatory |
| :----- | :---------- | :-------- |
| name | button's name | no |
| text | button's text | no |
| icon | button's Font Awesome icon | no |
| icon_right | if the icon should be on the right of the button's text instead of on the left | no |
| size | button's size (affects font, icon, height) | no |
| hover | button's hover text | no |
| onClick | function called on button click, receives the button object as an argument | no |

#### Attributes

| Attribute | Description |
| :-------- | :---------- |
| button | Thranduil button object |
| icon_str | icon string (fa-icon_name) |

#### Methods

**setLoading():** if the button has an icon, then this will set the icon to a loading animation. Useful when a button triggers some action that takes some time to perform.

**unsetLoading():** sets the button icon back to normal.

---

### Checkbox

```lua
yui.Checkbox({text = 'Checkbox text', onClick = function(self) print(1) end})
```

#### Construction options

| Option | Description | Mandatory |
| :----- | :---------- | :-------- |
| name | checkbox's name | no |
| text | checkbox's text | yes |
| size | checkbox's size (affects font, icon, height) | no |
| onClick | function called on checkbox click, receives the checkbox object as an argument | no |

#### Attributes

| Attribute | Description |
| :-------- | :---------- |
| checkbox | Thranduil checkbox object |
| checked | if the checkbox is checked or not |
| icon | checkbox checked icon, defaults to `yui.Theme.font_awesome['fa-check']` |

---

### Dropdown

```lua
yui.Dropdown({options = {'All', 'Option1', 'Option2'}, onSelect = function(self, option) print(option) end})
```

#### Construction options

| Option | Description | Mandatory |
| :----- | :---------- | :-------- |
| name | dropdown's name | no |
| options (table of strings) | dropdown options | yes |
| title | dropdown title (text to the left of it) | no |
| drop_up (boolean) | if the dropdown should go up instead of down | no |
| size | dropdown's size (affects font, icon, height) | no |
| onSelect | function called on option select, receives the object and the option as arguments | no |

#### Attributes

| Attribute | Description |
| :-------- | :---------- |
| main_button | Thranduil button object, main dropdown button |
| down_area | Thranduil frame object, visible and updates when the dropdown list is enabled |
| current_option | currently selected option |
| icon | dropdown main button's icon, defaults to `yui.Theme.font_awesome['fa-sort-desc']`

---

### FlatDropdown

Exactly the same as a Dropdown, just looks different.

### FlatTextinput

Exactly the same as a Textinput, just looks different.

### HorizontalSeparator

```lua
yui.HorizontalSeparator({w = 100})
```

#### Construction options

| Option | Description | Mandatory |
| :----- | :---------- | :-------- |
| name | horizontal separator's name | no |
| w | the width of the line | yes |
| margin_left | left margin | no |
| margin_right | right margin | no |
| size | separator's size (affects height) | no |

---

### HorizontalSpacing

```lua
yui.HorizontalSpacing({w = 50})
```

#### Construction options

| Option | Description | Mandatory |
| :----- | :---------- | :-------- |
| name | horizontal spacing's name | no |
| w | the width of the spacing | yes |
| size | spacing's size (affects height) | no |

---

### IconButton

```lua
yui.IconButton({icon = 'fa-close', onClick = function(self) print(1) end})
```

#### Construction options

| Option | Description | Mandatory |
| :----- | :---------- | :-------- |
| name | icon button's name | no |
| icon | font awesome icon | yes |
| hover | icon button's hover text | no |
| size | icon button's size (affects font, icon, height) | yes |
| onClick | function called on icon button click, receives the icon button object as an argument | no |

#### Attributes

| Attribute | Description |
| :-------- | :---------- |
| button | Thranduil button object |

---

### ImageButton

```lua
yui.ImageButton({image = some_image, onClick = function(self) print(1) end})
```

#### Construction options

| Option | Description | Mandatory |
| :----- | :---------- | :-------- |
| name | image button's name | no |
| rounded_corners (boolean) | if the image button has rounded corners or not | no |
| image | image button's image (LÖVE image object) | yes |
| ix, iy | image button's image offset (from top-left) | no |
| overlayNew | function called on object creation, receives the image button object as an argument | no
| overlayUpdate | function called every update, receives the image button object and dt as an argument | no |
| overlay | function called every draw call, receives the image button object as an argument | no |
| onClick | function called on image button click, receives the image button object as an argument | no |

#### Attributes

| Attribute | Description |
| :-------- | :---------- |
| button | Thranduil button object |
| img | image button image |
| alpha | overlay alpha (goes to 255 as the user hovers, back to 0 when not hovering)|

---

### Tabs

```lua
yui.Tabs({tabs = {
    {text = 'Tab1', hover = 'Tab1', onClick = function(self) print(1) end},
    {text = 'Tab2', hover = 'Tab2', onClick = function(self) print(2) end},
    {text = 'Tab3', hover = 'Tab3', onClick = function(self) print(3) end},
}})
```

#### Construction options

| Option | Description | Mandatory |
| :----- | :---------- | :-------- |
| name | tabs' name | no |
| tabs (table of tabs) | tabs list | yes |
| size | tabs' size (affects font, icon, height) | no |

#### Tab construction options

| Option | Description | Mandatory |
| :----- | :---------- | :-------- |
| text | tab's text | yes |
| hover | tab's hover text | yes |
| onClick | function called on tab selection, receives the tab object as an argument | no |

#### Attributes

| Attribute | Description |
| :-------- | :---------- |
| buttons | tabs Thranduil button objects |
| selected_tab | currently selected tab |

---

### Textinput

```lua
yui.Textinput({onEnter = function(self, text) print(text) end})
```

#### Construction options

| Option | Description | Mandatory |
| :----- | :---------- | :-------- |
| name | textinput's name | no |
| size | textinput's size (affects font and height) | no |
| onEnter | function called on pressing enter, receives the textinput object and the text as arguments | no |

#### Attributes

| Attribute | Description |
| :-------- | :---------- |
| textarea | Thranduil textinput object |

#### Methods

**setText():** sets the textinput's text

**getText():** returns the textinput's text

---

### VerticalSeparator

```lua
yui.VerticalSeparator({h = 100})
```

#### Construction options

| Option | Description | Mandatory |
| :----- | :---------- | :-------- |
| name | vertical separator's name | no |
| h | the height of the line | yes |
| margin_top | top margin | no |
| margin_bottom | bottom margin | no |
| size | separator's size (affects width) | no |

---

### VerticalSpacing

```lua
yui.VerticalSpacing({h = 50})
```

#### Construction options

| Option | Description | Mandatory |
| :----- | :---------- | :-------- |
| name | vertical spacing's name | no |
| h | the height of the spacing | yes |
| size | spacing's size (affects width) | no |

---

## Theme

The default theme file can be found in `yaoui/YaouiTheme.lua`. Internally yaoui uses [Thranduil](https://github.com/adonaac/thranduil) to make everything work, and so the theming system is the same here as it is there. Basically it's one big file with functions for all UI objects. The theme also has a few helpful variables (the theme itself can be accessed via `yui.Theme`):

| Variable | Description |
| :----- | :---------- |
| colors | a table containing all colors used by all elements |
| font_awesome | the Font Awesome icon table, `font_awesome[fa-icon_name]` returns you the icon character which can only be drawn with the Font Awesome font |
| font_awesome_path | path to the Font Awesome font |
| open_sans_regular | path to the Open Sans Regular font |
| open_sans_light | path to the Open Sans Light font |
| open_sans_bold | path to the Open Sans Bold font |
| open_sans_semibold | path to the Open Sans Semibold font |

### Changing the theme

I'm not going to list all theme colors here since you can easily check them yourself. But the easiest way to change the look of the theme is by changing the colors around. For instance, the following code swaps blue and red components for each color:

```lua
for color_name, color in pairs(yui.Theme.colors) do
  yui.Theme.colors[color_name] = {color[3], color[2], color[1]}
end
```

![red](http://i.imgur.com/O1nAnjo.jpg)

The second way of changing the theme is by specifying the draw functions of each element yourself. This is harder because it requires you to have an understanding of the attributes of each UI element and how they work, but it's very doable. 

You should start by removing all `new` and `update` functions, since those are only used if you need something like tweens or transitions. Then you should clear out the draw functions (remove all their content) and start defining, for each object you wanna use, how they will be drawn. Use Thranduil's TestTheme and yaoui's YaouiTheme as reference points and go from there. It's also important to keep all draw functions defined, even if they're empty, otherwise the program won't run.

## LICENSE

You can do whatever you want with this. See the [LICENSE](https://github.com/adonaac/yaoui/blob/master/LICENSE).


CHANGELOG.md
v0.1.1

## CHANGELOG

### v0.1.1


#### Fixes

* Fixed loading issues and removed winapi references - [b83f752](https://github.com/adonaac/blade/commit/b83f7526132227b55083c131b5453c7fc0d9e6b3) (2015-08-21 10:40:42)

---

### v0.1.0

>**Main functionality and a number of UI elements done**

[**0b96979**](https://github.com/adonaac/blade/commit/0b969798160abcceb41fe86d2c59b0d085022482) (**2015-07-30 12:34:20**)

#### Additions

* Added UI base lib - [944f335](https://github.com/adonaac/blade/commit/944f3358ed9986636d6f8e06e8136d7cb3c10512) (2015-07-21 04:49:54)
* Added base library structure - [4ab30d6](https://github.com/adonaac/blade/commit/4ab30d670e3b056cf215f54f9ab011a301c7884f) (2015-07-23 13:52:53)
* Added View class - [4ab30d6](https://github.com/adonaac/blade/commit/4ab30d670e3b056cf215f54f9ab011a301c7884f) (2015-07-23 13:52:53)
* Added Stack class - [4ab30d6](https://github.com/adonaac/blade/commit/4ab30d670e3b056cf215f54f9ab011a301c7884f) (2015-07-23 13:52:53)
* Added Flow class - [4ab30d6](https://github.com/adonaac/blade/commit/4ab30d670e3b056cf215f54f9ab011a301c7884f) (2015-07-23 13:52:53)
* Added IconButton element - [4ab30d6](https://github.com/adonaac/blade/commit/4ab30d670e3b056cf215f54f9ab011a301c7884f) (2015-07-23 13:52:53)
* Added 'name' and index based object retrieval from the UI tree - [4ab30d6](https://github.com/adonaac/blade/commit/4ab30d670e3b056cf215f54f9ab011a301c7884f) (2015-07-23 13:52:53)
* Added UI element automatic placement under Stacks or Flows - [b72025a](https://github.com/adonaac/blade/commit/b72025ae6dfea0ddecfeeacb2ca9aa30fa67cd52) (2015-07-24 12:47:02)
* Added Stack and Flow building logic - [b72025a](https://github.com/adonaac/blade/commit/b72025ae6dfea0ddecfeeacb2ca9aa30fa67cd52) (2015-07-24 12:47:02)
* Added Stack and Flow mixing logic - [b72025a](https://github.com/adonaac/blade/commit/b72025ae6dfea0ddecfeeacb2ca9aa30fa67cd52) (2015-07-24 12:47:02)
* Added Button - [bede56a](https://github.com/adonaac/blade/commit/bede56ac0ee1ce35fdc8eb744977b83f2005b738) (2015-07-26 07:32:10)
* Added size options to Button - [eb4a5c2](https://github.com/adonaac/blade/commit/eb4a5c2649c67f7b223639adfbe20ee7a97fe456) (2015-07-26 11:08:02)
* Added Checkbox - [0b177bb](https://github.com/adonaac/blade/commit/0b177bb8ff4989b3f80dcaed0f57a3aac26d5985) (2015-07-26 12:14:12)
* Added Text - [88fddbd](https://github.com/adonaac/blade/commit/88fddbd0f1267ae7bf1b7ce9b54ecc48aaa7148d) (2015-07-26 16:55:58)
* Added HorizontalSeparator - [093e52a](https://github.com/adonaac/blade/commit/093e52a51c5056c5df77f18e4c48a4cb32e30382) (2015-07-26 17:33:11)
* Added Dropdown - [cc25a7c](https://github.com/adonaac/blade/commit/cc25a7c6b1bc11f23664b1775863893dc8f7c7fd) (2015-07-27 08:22:19)
* Added FlatDropdown - [0afdacf](https://github.com/adonaac/blade/commit/0afdacffe6da10cc7088bd1dd68bd3043e77266a) (2015-07-27 11:12:05)
* Added onSelect callback for Dropdown and FlatDropdown - [0afdacf](https://github.com/adonaac/blade/commit/0afdacffe6da10cc7088bd1dd68bd3043e77266a) (2015-07-27 11:12:05)
* Added Textinput - [c375237](https://github.com/adonaac/blade/commit/c375237aea75e956f1494d45221d4618d03e9660) (2015-07-27 13:38:38)
* Added FlatTextinput - [5e00089](https://github.com/adonaac/blade/commit/5e000896fb13fcce2b8150e7019ee7f312c9b44c) (2015-07-27 21:28:17)
* Added a bunch of new settings to some elements - [9c83fd3](https://github.com/adonaac/blade/commit/9c83fd3667097452cdba838fa54c3ec424a3a880) (2015-07-28 05:43:37)
* Added onEnter option for Textinput and FlatTextinput - [4a80969](https://github.com/adonaac/blade/commit/4a8096925ffddb225865fcd83ade7a4d458f935c) (2015-07-28 06:02:46)
* Added ImageButton - [b98a646](https://github.com/adonaac/blade/commit/b98a646f30eee0eafd55f3c5602ce6dc2a02c4d9) (2015-07-28 07:36:47)
* Added onWidthChange callback to Dropdown - [b98a646](https://github.com/adonaac/blade/commit/b98a646f30eee0eafd55f3c5602ce6dc2a02c4d9) (2015-07-28 07:36:47)
* Added hover options for Button and IconButton - [035c254](https://github.com/adonaac/blade/commit/035c254b0e97646b8ca544e4384a7cf8ab23504d) (2015-07-28 09:56:33)
* Added winapi for opening dialogs - [9a5f8fb](https://github.com/adonaac/blade/commit/9a5f8fb9328c2759ec5d97e1db1c957d5c8b3943) (2015-07-28 11:09:22)
* Added openColorPicker call - [9f50bd6](https://github.com/adonaac/blade/commit/9f50bd6347fac80a0e858dd8b6e3e0864c7e28fe) (2015-07-28 12:10:47)
* Added openSaveDialog call - [aa2f91b](https://github.com/adonaac/blade/commit/aa2f91b991e9f41225cb8bb25d9c4f972e3d74ce) (2015-07-28 20:49:18)
* Added openOpenDialog call - [aaeb1eb](https://github.com/adonaac/blade/commit/aaeb1eb06511f11048bf25f28a35ce6385f77049) (2015-07-28 20:51:03)
* Added Tabs - [840e791](https://github.com/adonaac/blade/commit/840e791a61cb97d92e415286f176dc40909cf148) (2015-07-28 23:43:18)
* Added VerticalSeparator - [0dd9d68](https://github.com/adonaac/blade/commit/0dd9d685f1ab93a11c68799745b0d9214b1a8898) (2015-07-29 01:01:17)
* Added VerticalSpacing - [0dd9d68](https://github.com/adonaac/blade/commit/0dd9d685f1ab93a11c68799745b0d9214b1a8898) (2015-07-29 01:01:17)
* Added HorizontalSpacing - [0dd9d68](https://github.com/adonaac/blade/commit/0dd9d685f1ab93a11c68799745b0d9214b1a8898) (2015-07-29 01:01:17)
* Added colors table to YaouiTheme - [e815458](https://github.com/adonaac/blade/commit/e8154580d9c883f105a8147a64fd20334307f7b5) (2015-07-29 01:42:25)
* Added Anime example - [6564c3d](https://github.com/adonaac/blade/commit/6564c3d031dd826b7ee7c8f56b09e13c5cffbfd6) (2015-07-30 00:09:56)
* Added README - [71ee091](https://github.com/adonaac/blade/commit/71ee09124dd148842eeac2a1b652d520c075f297) (2015-07-30 11:25:34)

#### Fixes

* Fixed issues with Stack, Flow stacking - [e9f4e10](https://github.com/adonaac/blade/commit/e9f4e10a1e2dd95f47c63d1079c46b3032f46220) (2015-07-25 01:34:01)

yaoui/init.lua

local yaoui_path = ... .. '.'

local yaoui = {}
yaoui.UI = require(yaoui_path .. 'UI')
yaoui.Theme = require(yaoui_path .. 'YaouiTheme')
yaoui.Timer = require(yaoui_path .. 'Timer')

local View = require(yaoui_path .. 'View')
yaoui.View = function(...) return View(yaoui, ...) end
local Stack = require(yaoui_path .. 'Stack')
yaoui.Stack = function(...) return Stack(yaoui, ...) end
local Flow = require(yaoui_path .. 'Flow')
yaoui.Flow = function(...) return Flow(yaoui, ...) end

local Button = require(yaoui_path .. 'Button')
yaoui.Button = function(...) return Button(yaoui, ...) end
local Checkbox = require(yaoui_path .. 'Checkbox')
yaoui.Checkbox = function(...) return Checkbox(yaoui, ...) end
local Dropdown = require(yaoui_path .. 'Dropdown')
yaoui.Dropdown = function(...) return Dropdown(yaoui, ...) end
local FlatDropdown = require(yaoui_path .. 'FlatDropdown')
yaoui.FlatDropdown = function(...) return FlatDropdown(yaoui, ...) end
local FlatTextinput = require(yaoui_path .. 'FlatTextinput')
yaoui.FlatTextinput = function(...) return FlatTextinput(yaoui, ...) end
local HorizontalSeparator = require(yaoui_path .. 'HorizontalSeparator')
yaoui.HorizontalSeparator = function(...) return HorizontalSeparator(yaoui, ...) end
local HorizontalSpacing = require(yaoui_path .. 'HorizontalSpacing')
yaoui.HorizontalSpacing = function(...) return HorizontalSpacing(yaoui, ...) end
local IconButton = require(yaoui_path .. 'IconButton')
yaoui.IconButton = function(...) return IconButton(yaoui, ...) end
local ImageButton = require(yaoui_path .. 'ImageButton')
yaoui.ImageButton = function(...) return ImageButton(yaoui, ...) end
local Tabs = require(yaoui_path .. 'Tabs')
yaoui.Tabs = function(...) return Tabs(yaoui, ...) end
local Text = require(yaoui_path .. 'Text')
yaoui.Text = function(...) return Text(yaoui, ...) end
local Textinput = require(yaoui_path .. 'Textinput')
yaoui.Textinput = function(...) return Textinput(yaoui, ...) end
local VerticalSeparator = require(yaoui_path .. 'VerticalSeparator')
yaoui.VerticalSeparator = function(...) return VerticalSeparator(yaoui, ...) end
local VerticalSpacing = require(yaoui_path .. 'VerticalSpacing')
yaoui.VerticalSpacing = function(...) return VerticalSpacing(yaoui, ...) end

yaoui.update = function(views)
    love.mouse.setCursor()

    -- When a dropdown list is being interacted with, then prevent elements under it from being updated
    local all_elements = {}
    for _, view in ipairs(views) do
        local elements = view:getAllElements()
        for _, e in ipairs(elements) do table.insert(all_elements, e) end
    end
    local any_dropdown_area_hot = false
    for i, element in ipairs(all_elements) do
        element.dont_update = false
        if element.class_name == 'Dropdown' or element.class_name == 'FlatDropdown' then
            if element.down_area.hot then any_dropdown_area_hot = i end
        end
    end
    if any_dropdown_area_hot then
        for i, element in ipairs(all_elements) do
            if i == any_dropdown_area_hot then element.dont_update = false
            else element.dont_update = true end
        end
    end
end

yaoui.winapi = require(yaoui_path .. 'dialogs.winapi')
yaoui.colorchooser = require(yaoui_path .. 'dialogs.winapi.colorchooser')
yaoui.filedialogs = require(yaoui_path .. 'dialogs.winapi.filedialogs')

yaoui.openColorPicker = function()
    local cc, cust = yaoui.winapi.CHOOSECOLOR({flags = 'CC_FULLOPEN'})
    cc = yaoui.winapi.ChooseColor(cc)
    local bit = require('bit')
    local r, g, b = bit.band(bit.rshift(cc.result, 0), 255), bit.band(bit.rshift(cc.result, 8), 255), bit.band(bit.rshift(cc.result, 16), 255)
    return r, g, b
end

yaoui.openSaveDialog = function(title, filter, filter_index, flags)
    local ok, info = yaoui.winapi.GetSaveFileName({
        title = title,
        filter = filter,
        filter_index = filter_index,
        flags = flags or 'OFN_ALLOWMULTISELECT|OFN_EXPLORER|OFN_ENABLESIZING|OFN_FORCESHOWHIDDEN'
    })
    if ok then return info.filepath, info.filename, info.filter_index end
end

yaoui.openOpenDialog = function(title, filter, filter_index, flags)
    local ok, info = yaoui.winapi.GetOpenFileName({
        title = title,
        filter = filter,
        filter_index = filter_index,
        flags = flags or 'OFN_ALLOWMULTISELECT|OFN_EXPLORER|OFN_ENABLESIZING|OFN_FORCESHOWHIDDEN'
    })
    if ok then return info.filepath, info.filename, info.filter_index end
end

return yaoui
